### ISSUE TITLE
[Issue]: ongoing issues with styles

### ISSUE THREAD

**Issue by myndxero** [vladmandic/automatic#2629]
### Issue Description

I am using the right-click context menu only when encountering these issues. Some things I can't reproduce but will list here anyway:

The prompt that's saved is the one used in last generation only, this means it won't save the wildcards used, but the values chosen from them.
- would prefer and recommend old behavior of simply saving the current prompt in the text box

New entries will not show up in the styles dropdown menu even after refreshing, only works after restarting SD next. Doesn't appear in network styles section either.
- again would prefer old behavior and having it readily available after save without needing to restart SD next.

Maybe not an issue necessarily, but as an option would be nice: somehow someway, when I save an updated version of the prompt via the context menu, erases the previous one?
- unsure if bug or intended, seems more likely a bug, but optional intended behavior would be nice.

Saving via context sometimes results in an empty json entry. Has the fields listed but no data.
- it's possible this happens when saving a prompt, in its current form, while generation is occuring.
-  may also occur when SD next crashes/is improperly shutdown, before restarting it after saving via context menu.

Quality of life:
- would like a way to toggle or expunge the built-in styles from appearing anywhere they show up: styles dropdown, networks, and in particular, using the 'add all' via randomize extension and xyz grid.
-  expanding on this idea, would be nice to add a simple way to filter 'add all' contents, like typing the start of the name. For instance: since I keep all my models and styles organized, typing ____sdxl and selecting all under that for xyz grid would allow me to grab only my SDXL models and not include 2.X or 1.5 models. Same thing for styles.

My point of view has always been a desire for the old work flow be as simply and convenient as it was before the rework. I don't mind the compromise with your vision to cut back on buttons everywhere, but some functionality still needs fine-tuning.

Sorry I haven't been able to do more with everything and appreciate the patience for my ever-changing circumstances.

### Backend

Original

### Branch

Master

### Model

Other

### Acknowledgements

- [X] I have read the above and searched for existing issues
- [X] I confirm that this is classified correctly and its not an extension issue

**Comment by myndxero** [comment]
Just to add since it caught my eye, I also use prompt-all-in-one extension. It's possible, tho I personally don't see how, there could be a compatibility conflict here? I could look into that specifically maybe if warranted.

**Comment by vladmandic** [comment]
this is a good write-up and i'll provide response item-by-item.
but for any follow-ups, we need to separate issues - i'm going to close this one as "meta issue" is not something i can work with.

> The prompt that's saved is the one used in last generation only, this means it won't save the wildcards used, but the values chosen from them.

this is a question of preference. i prefer to have actual real prompt as saving wildcards makes that prompt dependent on wildcards - its unsuable unless something provides those wildcards. which means that style suddenly has an external dependency.

but i can see the use case, i may add option in settings to choose between save real prompt vs save it as-is.
in either case, that's a feature request - and i suggest you create one specific for that so it doesn't fall off the radar.

> New entries will not show up in the styles dropdown menu even after refreshing, only works after restarting SD next. Doesn't appear in network styles section either.

this has been fixed in dev branch.

> Maybe not an issue necessarily, but as an option would be nice: somehow someway, when I save an updated version of the prompt via the context menu, erases the previous one?

this has been fixed in dev branch.

> Saving via context sometimes results in an empty json entry. Has the fields listed but no data.

this has been fixed in dev branch.

> would like a way to toggle or expunge the built-in styles from appearing anywhere they show up: styles dropdown, networks, and in particular, using the 'add all' via randomize extension and xyz grid.

there is already an option in settings to disable built-in styles.

> expanding on this idea, would be nice to add a simple way to filter 'add all' contents, like typing the start of the name. For instance: since I keep all my models and styles organized, typing ____sdxl and selecting all under that for xyz grid would allow me to grab only my SDXL models and not include 2.X or 1.5 models. Same thing for styles.

this is not specific to styles, it would be a cool thing for xyz grid in general - but its non-trivial work and i don't see it as a priority at the moment. PRs are welcome!

**Comment by myndxero** [comment]
> this is a question of preference. i prefer to have actual real prompt as saving wildcards makes that prompt dependent on wildcards - its unsuable unless something provides those wildcards. which means that style suddenly has an external dependency.

I wonder if we could come up with enough reasons to separate styles and actual prompt saving. I've always used styles as a means to easily save and store my prompts, so I'm sort of left out in the cold here. That's fine as I'm ultimately not using styles _as intended_.

I like how the prompt all-in-one extension functions mostly, but the way you bookmark and save prompts with it gets clunky. It's nice both having the prompt history and, over time, builds a very nice comprehensive list of prompt keywords. Would any of that functionality possibly be of interest building into SD.next? Styles maintains it's place and this possibly offers some ideas into building something specific to prompt saving and a history but improving on how those two actually function vs the existing extension?

I've not set up my dev branch scenario for remote use yet, sorry for a long write up on things you've already handled largely. Expect these kinds of write-ups, if I could have the time to stay focused on dev branch scenarios and control net without other things getting in the way with my situation. Blah.

**Comment by vladmandic** [comment]
> Would any of that functionality possibly be of interest building into SD.next

you'll have to elaborate a bit more on this? is this about having a MRU list of prompts easily accesible?

**Comment by myndxero** [comment]
> you'll have to elaborate a bit more on this? is this about having a MRU list of prompts easily accesible?

Yes-ish.
![paio_01](https://github.com/vladmandic/automatic/assets/113237671/08b32f1f-ad1b-4baf-a4ba-32b654c941de)
![paio_02](https://github.com/vladmandic/automatic/assets/113237671/76bd7d36-3b99-4468-ae43-92d8a12ffbcd)

I like a lot of the concepts here provided by this extension, but not a big fan of the layout nor ease-of-use and presentation. This may need some brainstorming to keep any development here inline with your design philosophy.

What is most compelling to me:
- prompt history of course with a way to preview quickly
- keywords, the built-in is prob useful to others, potentially especially to new users, but it also keeps a history of all the keywords you use and allows add-in them point-click style
- biggest one for me is naming and saving prompts from the history
- there's advanced auto-formatting options and such too
- hide/expand arrows inline with the rest of how sd.next is presented to users
- saves prompt history for txt2txt/txt2img and negative prompts independently but all are accessible in each use-case

What I don't like:
- presentation overall
- prompt history is clunky to interact with
- the favorites tab fills up quickly and gets clunkier and clunkier to use
- it's a bit advanced to get used to and use overall, especially if just starting out

What I'd want to see built into sd.next:
- prompt history would be neat
- save prompt functionality in relation to how I preferred to use styles, allowing the saving of wildcards (you mentioned you didn't like that since reliant on external, and I take your meaning here that not everyone will have the same wildcard library, but dynamic prompts itself is standard with sd.next, just a consideration)
- easy/fast saving/editing of prompts

Predicted concerns:
- not dissimilar enough from styles
- not enough added functionality to justify development
- based on existing extension

Layout thoughts:
- prompt history hidden via hide/expand arrow. better way to present history other than tooltip style overlay?
- saved prompts dropbox could replace styles dropbox, but styles remains in networks? I don't know how much styles is used and if that would be a big inconvience to styles users, could be useful to have styles and prompts selectable there but potentially in eye-sore for your design philosophy haha
- keywords scenario is nice, might be worth having some quick access to them in the hide/expand arrow and then have a more comprehensive list plus built-in ones in the networks tab?
- previews functionality similar styles, and presented the same way, in the networks section?

I realize prompts and styles have a lot of overlap, but there's also enough of a difference, I think, in their use WHEN used as intended, may even consider creating a way for prompt saves to have some sort of built in wildcard list from wildcards used? Could probably parse a wildcard file and present it in a separate section of the json, or directly in the prompt reformatted to { | | | } etc.

** Might even consider a potential shortcut to just allowing wildcards again in styles saves with a means to parse and store wildcards there too and having a way to call them in the prompt when 'applied' in prompt box.

**Comment by vladmandic** [comment]
regarding wildcards in save styles (using prompt as-is, not after generate) is something i've asked you create a separate feature request for - i'll consider adding it as option (not as a replacement for current behavior).

regarding everything else, seems like this extension is quite feature rich and i don't think it would be a best use of limited time trying to reimplement stuff that already works. i might consider adding lightweight prompt history implemented as MRU list.

**Comment by Woisek** [comment]
@vladmandic My opinion on the wildcards template in the metadata: YES please!
Since I use wildcards, it is saved in the image under "Template" or the like. What I really love about this, is the fact, that this also serves as a kind of "backup" for the wildcard prompts in itself. It's not uncommon (for me), to get often get back to an image weeks or even months after generating and look up the prompt in the metadata. And then it's more than convenient to see the wildcard template in it, because, maybe I forget how I did it or even if I did it altogether.

So, for example, seeing the final prompt:
[Women in Science:9999]RAW full color photo portrait of a young woman ﻿Samantha Bokack with a black scientists coat, grey lob hair, (looking away from camera:1.4), (smiling:0.3), elaborate, (retro science fiction:1.3), (spaceship interior:1.4), intricate environment, (medium shot:1.4), (engineering room:1.2), (interface terminals, wires, displays, conduits:1.3), detailed skin, pores, wrinkles, hard eyes, soft lighting, realistic, hard shadow, (washed out color:1.2), intricate, high detail, analog style, (soft focus:0.6), film grain

And then the template (somewhere) in the metadata:
[Women in Science:9999]RAW full color photo portrait of a young woman __women_fullnames__ with a {white|blue|grey|black} scientists coat, __haircolors__ __women_hairstyle__, (looking away from camera:1.4), (smiling:0.3), elaborate, (retro science fiction:1.3), (spaceship interior:1.4), intricate environment, (medium shot:1.4), (engineering room:1.2), (interface terminals, wires, displays, conduits:1.3), detailed skin, pores, wrinkles, hard eyes, soft lighting, realistic, hard shadow, (washed out color:1.2), intricate, high detail, analog style, (soft focus:0.6), film grain

So, if I need/want to generate more images from that, I just copy the wildcard template and there is no need to tediously re-do the wildcard setup (if I don't see that I used a template and still remember, that I did 🤪).
In short, I don't see any reason why to _not_ save the wildcard template in the metadata or, de facto, in the generated txt files, if checked. So yes, please keep this in one or the other way available. 👍

**Comment by vladmandic** [comment]
@Woisek it would have to be OR, not AND.
metadata has limited size and saving both processed and unprocessed prompts gets really big. also complicates parsing by a lot.
yes, there is "save txt next to file" and txt file does not have size limits, but that txt file is dump of the same metadata thats inside jpeg/png image and i'm not going to have completely seprate handler just for that.

but discussion here was not about saving metadata at all, it was about do we use prompt before or after processing when doing style quick-save.
